specfile/macro_options.py

```diff
@@ -0,0 +1,523 @@
+        self.defaults = defaults.copy() if defaults is not None else {}
+
+    def __repr__(self) -> str:
+        tokens = repr(self._tokens)
+        defaults = repr(self.defaults)
+        return f"MacroOptions({tokens}, '{self.optstring}', {defaults})"
+
+    def __str__(self) -> str:
+        return "".join(str(t) for t in self._tokens)
+
+    def _valid_option(self, name: str) -> bool:
+        """
+        Determines if a name represents a recognized option.
+
+        Args:
+            name: Name of the option.
+
+        Returns:
+            True if the option is recognized, otherwise False.
+        """
+        try:
+            # use parent's __getattribute__() so this method can be called from __getattr__()
+            optstring = super().__getattribute__("optstring")
+        except AttributeError:
+            return False
+        return name in optstring.replace(":", "")
+
+    def _requires_argument(self, option: str) -> bool:
+        """
+        Determines if an option requires an argument.
+
+        Args:
+            option: Name of the option.
+
+        Returns:
+            True if the option requires an argument, otherwise False.
+        """
+        i = self.optstring.index(option) + 1
+        return i < len(self.optstring) and self.optstring[i] == ":"
+
+    def _find_option(self, name: str) -> Tuple[int, int]:
+        """
+        Searches for an option in tokens of an option string.
+
+        Args:
+            name: Name of the option.
+
+        Returns:
+            Tuple of indices where the first is the index of a token matching
+            the option and the second is the index of a token matching
+            its argument, or -1 if there is no match.
+        """
+        option = f"-{name}"
+        for i, token in reversed(list(enumerate(self._tokens))):
+            if not token.value.startswith(option):
+                continue
+            if token.value != option:
+                return i, i
+            if not self._requires_argument(name):
+                return i, -1
+            j = i + 1
+            if j == len(self._tokens):
+                return i, -1
+            if self._tokens[j].type == TokenType.WHITESPACE:
+                j += 1
+            if j == len(self._tokens):
+                return i, -1
+            if self._tokens[j].value.startswith("-"):
+                return i, -1
+            return i, j
+        return -1, -1
+
+    @staticmethod
+    def _needs_quoting(value):
+        # if there is a whitespace, enquote the value rather than escaping it
+        return any(ws in str(value) for ws in string.whitespace)
+
+    def __getattr__(self, name: str) -> Union[bool, int, str]:
+        if not self._valid_option(name):
+            return super().__getattribute__(name)
+        i, j = self._find_option(name)
+        if i < 0:
+            if self._requires_argument(name):
+                return self.defaults.get(name)
+            return False
```

Comment:
Shouldn't getattr raise `AttributeError` if no match is found?

Reply:
Yes, but in this case it's a valid flag option, for example `-q` of `%setup`, if it's not present it's like it was set to `False`.

In my opinion it makes more sense to write `if prep.setup.options.q:` rather than `if q in prep.setup.options:` (or catching `AttributeError`).

Or you can think of it as if all flag options had `False` as default value.